Add additional encryption config flags + labels

ReToCode opened the pull request:
# Changes
- In preparation for https://github.com/knative/serving/issues/14216
- Add `CertificateTypeLabelKey` label with values `CertificateType` to indicate different types of a `KnativeCertificate`
- Drop deprecated and now unused `ServingInternalCertName` (checked in all repos in knative + knative-extensions)
- Add helper method `GetIngressTLSForVisibility` to `Ingress` to filter cluster-local and external-ip TLS blocks.

/kind enhancement

Example of `KnativeIngress` with TLS block for cluster-local-domain-tls:

```yaml
apiVersion: networking.internal.knative.dev/v1alpha1
kind: Ingress
metadata:
  name: helloworld
  namespace: default
spec:
  httpOption: Enabled
  rules:
  - hosts:
    - helloworld.default
    - helloworld.default.svc
    - helloworld.default.svc.cluster.local
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ClusterLocal
  - hosts:
    - helloworld.default.10.89.0.200.sslip.io
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
  tls:
  - hosts:
    - helloworld.default
    - helloworld.default.svc
    - helloworld.default.svc.cluster.local
    secretName: route-ba07ea4e-2548-4632-9187-e615c876cffc-local
    secretNamespace: default
```

ReToCode commented on pkg/apis/networking/v1alpha1/ingress_types.go:
```diff
@@ -149,6 +149,10 @@ type IngressTLS struct {
 	SecretNamespace string `json:"secretNamespace,omitempty"`
+
+	// Visibility signifies whether the tls hosts should be considered `ClusterLocal`.
+	// If it's not specified then it defaults to `ExternalIP`.
```
Done

nak3 commented:
/lgtm
/hold

Thank you!
For `knative-extensions/net-istio`, I think we need to add `IngressTLS.Vibility` in the unit test.
`/hold` for other reviewers. Please feel free to unhold if no one will take a look.

ReToCode commented:
> For knative-extensions/net-istio, I think we need to add IngressTLS.Vibility in the unit test.

👍 I'll do that as a follow-up

dprotaso commented:
> Add visibility to KnativeIngress.TLS (see example below)

Why does the ingress need this?

ReToCode commented:
> Why does the ingress need this?

Hm, thinking about it based on your question, it is probably not totally necessary. But without having it, each net-* solution would need to compare the `tls` section with the `hosts` list to find out if the tls config is relevant for a cluster-local domain or an external domain. In net-kourier, we use this flag to distinguish which listener to configure:

-  https://github.com/knative-extensions/net-kourier/pull/1156/files#diff-ba0933c476bd2d5262d8a049d818bfdb85e4367bd9fe4e44e6344e59d3245745R116
- https://github.com/knative-extensions/net-kourier/pull/1156/files#diff-e42194f399f5b50c7552aef224c002778531fdf92f11ae4be81e85ba6968b0a9R73

nak3 commented:
Should we restructure the `spec.tls` and `spec.rules` at this time? For example,

<details>
<summary> option a) merge `spec.tls` into `spec.rules` like:</summary>

```yaml
apiVersion: networking.internal.knative.dev/v1alpha1
kind: Ingress
metadata:
  name: helloworld
  namespace: default
spec:
  httpOption: Enabled
  rules:
  - hosts:
    - helloworld.default
    - helloworld.default.svc
    - helloworld.default.svc.cluster.local
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ClusterLocal
    secretName: some-external-secret  # optional - omitting means no TLS.
    secretNamespace: default
  - hosts:
    - helloworld.default.10.89.0.200.sslip.io
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ExternalIP
    secretName: route-ba07ea4e-2548-4632-9187-e615c876cffc-local # optional - omitting means no TLS.
    secretNamespace: default
```

</details>

<details>
<summary>option b) drop `spec.tls.hosts[]` and use the `spec.tls.visibility` as a key like:</summary>


```yaml
apiVersion: networking.internal.knative.dev/v1alpha1
kind: Ingress
metadata:
  name: helloworld
  namespace: default
spec:
  httpOption: Enabled
  rules:
  - hosts:
    - helloworld.default
    - helloworld.default.svc
    - helloworld.default.svc.cluster.local
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ClusterLocal
  - hosts:
    - helloworld.default.10.89.0.200.sslip.io
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ExternalIP
  tls:
  - visibility: ExternalIP
    secretName: secret-for-external
    secretNamespace: default
  - visibility: ClusterLocal
    secretName: route-ba07ea4e-2548-4632-9187-e615c876cffc-local
    secretNamespace: default
```

</details>

Also, `httpOption` should be applied to local and external respectively.

ReToCode commented:
Hm, not sure if that works. Note we can have multiple entries for both visibilities with different certificates:
Here an example with tags:

```yaml
apiVersion: networking.internal.knative.dev/v1alpha1
kind: Ingress
metadata:
  labels:
    serving.knative.dev/route: helloworld
    serving.knative.dev/routeNamespace: default
    serving.knative.dev/service: helloworld
  name: helloworld
  namespace: default
spec:
  httpOption: Enabled
  rules:
  - hosts:
    - helloworld.default
    - helloworld.default.svc
    - helloworld.default.svc.cluster.local
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ClusterLocal
  - hosts:
    - helloworld.default.192.168.105.100.sslip.io
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ExternalIP
  - hosts:
    - latest-helloworld.default
    - latest-helloworld.default.svc
    - latest-helloworld.default.svc.cluster.local
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ClusterLocal
  - hosts:
    - latest-helloworld.default.192.168.105.100.sslip.io
    http:
      paths:
      - splits:
        - appendHeaders:
            Knative-Serving-Namespace: default
            Knative-Serving-Revision: helloworld-00001
          percent: 100
          serviceName: helloworld-00001
          serviceNamespace: default
          servicePort: 443
    visibility: ExternalIP
  tls:
  - hosts:
    - helloworld.default
    - helloworld.default.svc
    - helloworld.default.svc.cluster.local
    secretName: route-9f44405d-74a9-496b-a43d-effdf98bd6ca-local
    secretNamespace: default
    visibility: ClusterLocal
  - hosts:
    - latest-helloworld.default
    - latest-helloworld.default.svc
    - latest-helloworld.default.svc.cluster.local
    secretName: route-9f44405d-74a9-496b-a43d-effdf98bd6ca-147587726-local
    secretNamespace: default
    visibility: ClusterLocal
```

nak3 commented:
I just realized that another concern of ths PR. Current net-* implementaitons does a simple `loop` against `spec.TLS` or `len(ing.Spec.TLS) > 0`. If we added the cluster local TLS, it will break without their support of cluster local?
Do you have a plan how to manage it like filter it by `spec.TLS.Visiblity`?

* net-istio
https://github.com/knative-extensions/net-istio/blob/e07187e33415e9c04a7c7ce473c083791e870feb/pkg/reconciler/ingress/ingress.go#L422

* net-contour
https://github.com/knative-extensions/net-contour/blob/c5932aa6c79e79f3d482dfe2b8d8c87811eca9ce/pkg/reconciler/contour/resources/httpproxy.go#L107
https://github.com/knative-extensions/net-contour/blob/c5932aa6c79e79f3d482dfe2b8d8c87811eca9ce/pkg/reconciler/contour/resources/httpproxy.go#L107

ReToCode commented:
That is a good point @nak3 , but I think not one that we can omit without introducing a totally new field here. The flag is still marked as alpha, so I'd vote to add the hint as you say. As long as this feature is not finished, it should not be enabled in a real environment. Then the cluster-local TLS entries will not occur.

I'll add this filter changes as issues to our tracking umbrella.

dprotaso commented:
I think I'd like to avoid API changes to our internal Ingress resource - since we'd have to support multiple versions while we perform a migration.

Given that the TLS stanza just maps host names to a secret I think I'd rather just create helper functions in this repo that implementations can use to reduce complexity.

ReToCode commented:
/hold I'm doing some testing with kourier + serving and add a test for the new function.

skonto commented:
> Should we restructure the spec.tls and spec.rules at this time? For example,

I would vote for that too. It seems there is a lot of repetition of info here.

dprotaso commented:
> I would vote for that too. It seems there is a lot of repetition of info here.

You could argue that you get _more_ repetition if you have a TLS block within the rules block. If you didn't have wildcard certs then you'd have to repeat the entire rule many times (eg. appendHeaders)

skonto commented:
> You could argue that you get more repetition if you have a TLS block within the rules block. 

We could have it under rules, my thinking was optimize based on what is common in practice, otherwise I agree:

```
spec:
  rules:
  - hosts:
    - helloworld.default
    - helloworld.default.svc
    - helloworld.default.svc.cluster.local
 tls:   
 ```

ReToCode commented:
@skonto  @dprotaso I don't think we ever have a wildcard in `king`. It seems to be only in the `kcert`, see:
- [domain-mapping with tls](https://github.com/ReToCode/knative-encryption/blob/main/7-cert-manager-only/results/king-domainmapping-tls.yaml)
- [with namespace wildcards](https://github.com/ReToCode/knative-encryption/blob/main/7-cert-manager-only/results/king-wildcard.yaml)
- [kcert with wildcard](https://github.com/ReToCode/knative-encryption/blob/main/7-cert-manager-only/results/kcert-wildcard.yaml#L23)

But anyway, I added `GetIngressTLSForVisibility` on `Ingress` and updated the PR description above. With this there are no changes to the API, we only need to make sure the net-* repos use the new method before one enables `cluster-local-domain-tls`. Full test run with kourier, net-certmanager is in Serving, see [here](https://github.com/knative/serving/actions/runs/6958392442/job/18933452251). Can you take another look?

dprotaso commented:
Can you update the commit message - it doesn't match the contents of the commit

dprotaso commented on pkg/apis/networking/v1alpha1/ingress_helpers.go:
```diff
@@ -0,0 +1,47 @@
+		if r.Visibility == visibility {
+			for _, t := range i.Spec.TLS {
+				// Check if hosts slices are equal ignoring the order
+				if cmp.Diff(r.Hosts, t.Hosts, cmpopts.SortSlices(func(a, b string) bool { return a < b })) == "" {
```
This shouldn't be an equal comparison but `tls.Hosts.Contains(host)`

In theory KIngress can express different TLS certs for different hosts that are in the same rules block

ReToCode commented on pkg/apis/networking/register.go:
```diff
@@ -119,6 +119,10 @@ const (
+
+	// CertificateTypeLabelKey is the label to indicate the type of Knative certificate
+	// used for Knative Serving encryption functionality.
+	CertificateTypeLabelKey = PublicGroupName + "/certificate-type"
```
Added a comment to link to `config.CertificateType`. Is there a better way to do this?

dprotaso commented:
/lgtm
/approve
